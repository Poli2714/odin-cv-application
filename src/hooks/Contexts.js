import { createContext } from 'react';

const IndexContext = createContext({
  activeIndex: null,
  setActiveIndex: () => {},
});

const KeyContext = createContext({
  activeKey: {
    eduActiveKey: null,
    expActiveKey: null,
  },
  setActiveKey: () => {},
});

const EduErrMsgContext = createContext({
  showEduErrMessage: false,
  setShowEduErrMessage: () => {},
});

const ExpErrMsgContext = createContext({
  showExpErrMessage: false,
  setShowExpErrMessage: () => {},
});

const PersonalInfoContext = createContext({
  personalInfo: {
    profilePicture: '',
    fullName: '',
    currentJobTitle: '',
    email: '',
    mobile: '',
    city: '',
  },
  setPersonalInfo: () => {},
});

const SummaryContext = createContext({
  summary: '',
  setSummary: () => {},
});

const ExperienceContext = createContext({
  experiences: {
    id: {
      jobTitle: '',
      employer: '',
      jobStartDate: '',
      jobEndDate: '',
      jobDescription: '',
    },
  },
  setExperiences: () => {},
});

const EducationContext = createContext({
  educations: {
    id: {
      degree: '',
      university: '',
      eduStartDate: '',
      eduEndDate: '',
      eduInfo: '',
    },
  },
  setEducations: () => {},
});

export {
  EducationContext,
  EduErrMsgContext,
  ExperienceContext,
  ExpErrMsgContext,
  IndexContext,
  KeyContext,
  PersonalInfoContext,
  SummaryContext,
};
