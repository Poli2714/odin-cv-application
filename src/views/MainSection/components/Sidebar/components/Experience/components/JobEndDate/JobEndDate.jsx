import { useContext } from 'react';
import PropTypes from 'prop-types';

import { InputWrapper } from 'src/components/layout';
import { Label, TextInput } from 'src/components/forms';
import { formatDateInput, isValidEndDate } from 'src/utils';
import { ExperienceContext } from 'src/hooks/Contexts';

function JobEndDate({ id }) {
  const { experiences, setExperiences } = useContext(ExperienceContext);
  const { jobEndDate, jobStartDate } = experiences[id];

  const handleChangeJobEndDate = e => {
    if (!isValidEndDate(jobStartDate, e.target.value)) return;

    const formattedDate = formatDateInput(e.target.value, jobEndDate);
    if (formattedDate === jobEndDate) return;

    setExperiences({
      ...experiences,
      [id]: {
        ...experiences[id],
        jobEndDate: formattedDate,
      },
    });
  };

  return (
    <InputWrapper>
      <Label labelName="End" labelFor="jobEndDate" />
      <TextInput
        disabled={jobStartDate.length < 9}
        name="jobEndDate"
        onChange={handleChangeJobEndDate}
        placeholder="MM / YYYY"
        value={jobEndDate}
      />
    </InputWrapper>
  );
}

JobEndDate.propTypes = {
  id: PropTypes.string,
};

export default JobEndDate;
