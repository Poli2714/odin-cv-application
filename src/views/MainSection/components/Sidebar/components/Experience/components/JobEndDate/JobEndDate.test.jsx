import { expect, test, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useState } from 'react';
import PropTypes from 'prop-types';

import JobEndDate from './JobEndDate';
import { ExperienceContext } from 'src/hooks/Contexts';

vi.mock('src/components/layout', () => ({
  InputWrapper: vi.fn(({ children }) => (
    <div data-testid="input-wrapper">{children}</div>
  )),
}));

vi.mock('src/components/forms', () => ({
  Label: vi.fn(() => <label htmlFor="jobEndDate">Mock label</label>),
  TextInput: vi.fn(({ disabled, onChange, value }) => (
    <input
      disabled={disabled}
      id="jobEndDate"
      type="text"
      onChange={onChange}
      value={value}
    />
  )),
}));

function MockParentComponent({ id }) {
  const [experiences, setExperiences] = useState({
    testId: { jobEndDate: '', jobStartDate: '07 / 2020' },
    testId2: { jobEndDate: '03 / 2024', jobStartDate: '01 / 1999' },
    testId3: { jobEndDate: '', jobStartDate: '01' },
  });

  return (
    <ExperienceContext.Provider value={{ experiences, setExperiences }}>
      <JobEndDate id={id} />
    </ExperienceContext.Provider>
  );
}

MockParentComponent.propTypes = {
  id: PropTypes.string,
};

test('renders JobEndDate', () => {
  render(<MockParentComponent id="testId" />);
  const input = screen.getByLabelText('Mock label');

  expect(screen.getByTestId('input-wrapper')).toMatchSnapshot();
  expect(input).not.toBeDisabled();
  expect(input).toHaveValue('');
});

test('renders JobEndDate with different initial value', () => {
  render(<MockParentComponent id="testId2" />);
  const input = screen.getByLabelText('Mock label');

  expect(input).not.toBeDisabled();
  expect(input).toHaveValue('03 / 2024');
});

test('disables jobEndDate input if jobStartDate is empty/incomplete', async () => {
  const user = userEvent.setup();

  render(<MockParentComponent id="testId3" />);
  const input = screen.getByLabelText('Mock label');

  await user.click(input);
  expect(input).not.toHaveFocus();
  expect(input).toBeDisabled();
  await user.type(input, '12');
  expect(input).toHaveValue('');
});

test('updates jobEndDate value as a user types', async () => {
  const user = userEvent.setup();

  render(<MockParentComponent id="testId" />);
  const input = screen.getByLabelText('Mock label');

  await user.type(input, '072023');
  expect(input).toHaveValue('07 / 2023');

  await user.clear(input);
  await user.type(input, '12');
  expect(input).toHaveValue('12 / ');

  await user.clear(input);
  await user.type(input, '1020221');
  expect(input).toHaveValue('10 / 2022');
});

test("doesn't update jobEndDate value when a user enters an invalid number", async () => {
  const user = userEvent.setup();

  render(<MockParentComponent id="testId" />);
  const input = screen.getByLabelText('Mock label');

  await user.type(input, '2');
  expect(input).toHaveValue('');

  await user.type(input, ' ');
  expect(input).toHaveValue('');

  await user.type(input, '1021');
  expect(input).toHaveValue('10 / 2');

  await user.clear(input);
  await user.type(input, '2109475');
  expect(input).toHaveValue('10 / ');
});
