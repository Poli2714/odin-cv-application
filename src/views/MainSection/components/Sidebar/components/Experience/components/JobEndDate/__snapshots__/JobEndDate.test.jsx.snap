// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders JobEndDate 1`] = `
<div
  data-testid="input-wrapper"
>
  <label
    for="jobEndDate"
  >
    Mock label
  </label>
  <input
    id="jobEndDate"
    type="text"
    value=""
  />
</div>
`;
